comb_spec_searcher/isomorphism.py

```diff
@@ -0,0 +1,251 @@
+        self._ancestors1: Dict[CombinatorialClass, int] = {}
+        self._root2: CombinatorialClass = spec2.root
+        self._rules2: Dict[CombinatorialClass, AbstractRule] = spec2.rules_dict
+        self._ancestors2: Dict[CombinatorialClass, int] = {}
+        self._order_map: Dict[
+            Tuple[CombinatorialClass, CombinatorialClass], List[int]
+        ] = {}
+        self._isomorphic = self._are_isomorphic(self._root1, self._root2)
+
+    def are_isomorphic(self) -> bool:
+        """Check if the two specs are isomorphic."""
+        return self._isomorphic
+
+    def get_order(self, c1: CombinatorialClass, c2: CombinatorialClass) -> List[int]:
+        """Get order map of corresponding nodes."""
+        return self._order_map[(c1, c2)]
+
+    def _are_isomorphic(
+        self, node1: CombinatorialClass, node2: CombinatorialClass
+    ) -> bool:
+        # If there are equivilances, we use the 'latest' one
+        node1, node2 = Isomorphism._get_eq_descendant(
+            node1, self._rules1, node2, self._rules2
+        )
+        rule1, rule2 = self._rules1[node1], self._rules2[node2]
+
+        is_base_case = self._base_cases(node1, rule1, node2, rule2)
+        if is_base_case:
+            return bool(is_base_case + 1)
+
+        self._index += 1
+        self._ancestors1[node1], self._ancestors2[node2] = (self._index, self._index)
+
+        # Check all matches of children, if any are valid then trees are isomorphic
+        n = len(rule1.children)
+        child_order: List[int] = [0] * n
+        stack = [(0, i, {i}) for i in range(n - 1, -1, -1)]
+        while stack:
+            i1, i2, in_use = stack.pop()
+            if not self._are_isomorphic(rule1.children[i1], rule2.children[i2]):
+                continue
+            child_order[i1] = i2
+            if i1 == n - 1:
+                self._order_map[(node1, node2)] = child_order
+                self._cleanup(node1, node2)
+                return True
+            Isomorphism._extend_stack(i1, n, in_use, stack)
+
+        self._cleanup(node1, node2)
+        return False
+
+    @staticmethod
+    def _get_eq_descendant(
+        node1: CombinatorialClass,
+        rules1: Dict[CombinatorialClass, AbstractRule],
+        node2: CombinatorialClass,
+        rules2: Dict[CombinatorialClass, AbstractRule],
+    ) -> Tuple[CombinatorialClass, CombinatorialClass]:
+        rule1, rule2 = rules1[node1], rules2[node2]
+
+        if rule2.is_equivalence():
+            node2 = rule2.children[0]
+
+        if rule1.is_equivalence():
+            node1 = rule1.children[0]
+
+        return node1, node2
+
+    def _base_cases(
+        self,
+        node1: CombinatorialClass,
+        rule1: AbstractRule,
+        node2: CombinatorialClass,
+        rule2: AbstractRule,
+    ) -> int:
+        # If different type of rules are applied, the trees are not isomorphic
+        if rule1.strategy.get_op_symbol() != rule2.strategy.get_op_symbol():
```

Comment:
We should probably rely on the constructor rather than the op symbol. That is `rule.constructor`.